_posts/2018-08-04-pfsense-as-hvm-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,574 @@
+libxl: error: libxl_pci.c:409:libxl_device_pci_assignable_list: Looks like pciback driver not loaded
+libxl: error: libxl_pci.c:1225:libxl__device_pci_add: PCI device 0:2:0.0 is not assignable
+libxl: error: libxl_pci.c:1304:libxl__add_pcidevs: libxl_device_pci_add failed: -3
+libxl: error: libxl_create.c:1461:domcreate_attach_devices: unable to add pci devices
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 1
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 1
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 1 failed
+```
+
+Solution:
+
+```
+modprobe xen-pciback
+```
+
+# PCI device not assignable
+
+```
+libxl: error: libxl_pci.c:1225:libxl__device_pci_add: PCI device 0:2:0.0 is not assignable
+libxl: error: libxl_pci.c:1304:libxl__add_pcidevs: libxl_device_pci_add failed: -3
+libxl: error: libxl_create.c:1461:domcreate_attach_devices: unable to add pci devices
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 2
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 2
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 2 failed
+```
+
+Assign PCI device using `xl pci-assignable-add`.
+
+# No IOMMU
+
+```
+root@apu2:~# xl create pfsense.cfg
+Parsing config from pfsense.cfg
+libxl: error: libxl_pci.c:1209:libxl__device_pci_add: PCI device 0000:02:00.0 cannot be assigned - no IOMMU?
+libxl: error: libxl_pci.c:1304:libxl__add_pcidevs: libxl_device_pci_add failed: -1
+libxl: error: libxl_create.c:1461:domcreate_attach_devices: unable to add pci devices
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 9
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 9
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 9 failed
+```
+
+This error means you don't have IOMMU correctly enabled. For AMD platforms `xl
+dmesg` contain:
+
+```
+root@apu2:~# xl dmesg|grep -i iommu
+(XEN) AMD-Vi: IOMMU not found!
+```
+
+# Lack of block backend
+
+`xen-blkback` should be loaded or compiled in otherwise blow error pop-up.
+
+```
+root@apu2:~# xl create pfsense.cfg
+Parsing config from pfsense.cfg
+libxl: error: libxl_device.c:1086:device_backend_callback: unable to add device with path /local/domain/0/backend/vbd/1/51712
+libxl: error: libxl_create.c:1255:domcreate_launch_dm: unable to add disk devices
+libxl: error: libxl_device.c:1086:device_backend_callback: unable to remove device with path /local/domain/0/backend/vbd/1/51712
+libxl: error: libxl.c:1647:devices_destroy_cb: libxl__devices_destroy failed for 1
+libxl: error: libxl.c:1575:libxl__destroy_domid: non-existant domain 1
+libxl: error: libxl.c:1534:domain_destroy_callback: unable to destroy guest with domid 1
+libxl: error: libxl.c:1463:domain_destroy_cb: destruction of domain 1 failed
+```
+
+# Crash after couple tries
+
+After couple tries of creating pfSense VM I faced below error:
```

Comment:
After a couple of tries of creating pfSense VM I faced below error:

Reply:
done